### PR TITLE
Defensive statements before looking into VCs

### PR DESCRIPTION
**Pull request by garciafdezpatricia**
Code will now check if the VC provided is of the supported types, namely: DatasetWithId, VerifiableCredential, UrlString or URL. If it is not, it will be parsed into a valid VC before continuing with the execution flow, avoiding the code from breaking (due to the switch from JSON to JSON-LD VCs breaking change introduced in v3).

- [x] I've added a unit test to test for potential regressions of this bug.
- [ ] The changelog has been updated, if applicable.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

### CODE REVIEW

File: src/common/verify/isValidAccessGrant.ts [hunk 1]
```diff
@@ -28,8 +28,10 @@ import {
   getIssuer,
   getVerifiableCredentialApiConfiguration,
 } from "@inrupt/solid-client-vc";
+import { isUrl } from "@inrupt/solid-client-vc/dist/common/common";
```
Comment (NSeydoux) [inline src/common/verify/isValidAccessGrant.ts:31]:
```suggestion
import { isUrl } from "@inrupt/solid-client-vc/common";
```
This should be aligned with the `exports` in https://github.com/inrupt/solid-client-vc-js/blob/c18412db41f03aeab1024fef0b63b28bf7ba0bad/package.json#L34.

Comment (NSeydoux) [reply src/common/verify/isValidAccessGrant.ts:31]: Ah interesting, it doesn't resolve the issue.

**Comment by NSeydoux** [comment]
`isUrl` is probably something we'll prefer to locally re-implement (internally, and not exporting it in the public API) insteadu of importing it from the VC library.

**Comment by NSeydoux** [comment]
It looks rollup is also complaining about the subpath import, so there must be something wrong about it, my bad

**Comment by NSeydoux** [comment]
Our current rollup config is set up in a way that prevents this, but I think it would not be too problematic to change. I'll give it a go tomorrow.

File: src/common/verify/isValidAccessGrant.ts [hunk 2]
```diff
@@ -48,7 +51,20 @@ async function isValidAccessGrant(
...
+    typeof vc !== "string" &&
+    !isUrl(vc.toString()) &&
+    !(vc instanceof URL) &&
+    !isDatasetCore(vc)
+  ) {
+    validVC = await verifiableCredentialToDataset(vc, {
+      includeVcProperties: true,
+      requireId: false,
+    });
+  }
```
Comment (NSeydoux) [inline src/common/verify/isValidAccessGrant.ts:65]:
```suggestion
  if (
    typeof vc === "object" &&
    !isDatasetCore(vc) &&
    typeof (vc as { id: string }).id === "string"    
  ) {
    validVC = await verifiableCredentialToDataset(vc, {
      includeVcProperties: true,
      requireId: true,
    });
  }
```
I think I would prefer the logic here shifted to check whether the object that got passed is indeed a plain JSON representation of a Verifiable Credential, which would have a `.id` property. It is more restrictive than going by elimination, and assuming that a non-string non-URL non-Datasetcore object would be parseable as JSON-LD.

File: src/common/verify/isValidAccessGrant.ts [hunk 3]
```diff
@@ -48,7 +51,20 @@ async function isValidAccessGrant(
...
+  // validVc is null if the provided VC is of the supported types
+  const vcObject = await getBaseAccess(validVC ?? vc, options);
```
Comment (NSeydoux) [inline src/common/verify/isValidAccessGrant.ts:67]:
```suggestion
  if(validVc === null && isUrl(vc.toString())) {
    validVc = vc;
  }
  if(validVc === null) {
    throw new Error(`Invalid argument: expected either a VC URL or a RDFJS DatasetCore, received ${vc}`)
  }
  
  const vcObject = await getBaseAccess(validVc, options);
```
This allows to add an explicit error being thrown in case we didn't know how to handle the provided argument.

File: src/common/verify/isValidAccessGrant.ts [hunk 4]
```diff
@@ -48,7 +51,20 @@ async function isValidAccessGrant(
   } = {},
 ): Promise<{ checks: string[]; warnings: string[]; errors: string[] }> {
   const fetcher = await getSessionFetch(options);
-  const vcObject = await getBaseAccess(vc, options);
+  let validVC = null;
```
Comment (NSeydoux) [inline src/common/verify/isValidAccessGrant.ts:54]:
```suggestion
  let validVc = null;
```
That's minor, but introduces a naming inconsistency otherwise.

File: src/gConsent/guard/isDatasetCore.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ * @returns true if the candidate implements the DatasetCore interface, false otherwise.
+ * @internal
+ */
+function isDatasetCore(
+  candidateDataset: unknown | DatasetCore,
+): candidateDataset is DatasetCore {
+  return typeof (candidateDataset as DatasetCore).add === "function";
+}
+
+export { isDatasetCore };
```
Comment (NSeydoux) [inline src/gConsent/guard/isDatasetCore.ts:37]:
```suggestion
export function isDatasetCore(
  candidateDataset: unknown | DatasetCore,
): candidateDataset is DatasetCore {
  return typeof (candidateDataset as DatasetCore).add === "function";
}
```
The export declarations are inlined in the rest of the codebase.

**Review by NSeydoux — CHANGES_REQUESTED** [review]
Can you extract out the validation code into a shared `toVcDataset` that implements both the logic you have here (converting a JSON object into a `DatasetCore`) and lifts the dereferencing from `getBasedAccess`, i.e. factors out
```
if (typeof vc === "string" || vc instanceof URL) {
    baseVc = await getVerifiableCredential(vc.toString(), {
      returnLegacyJsonld: false,
      skipValidation: true,
      fetch: options.fetch,
    });
  }
``` 
? This way, `getBaseAccess` can be simplified to only take a `DatasetWithId` in, and `toVcDataset` is the single function that validates the input VC, either by fetching it or by converting it from plain JS object to `DatasetCore`. you can then reuse that `toVcDataset`
functions in each place you've implemented the validation logic in this PR.

Does that make sense?

**Comment by NSeydoux** [comment]
Also, can you add a unit test to verify that converting from a plain JS object to a `DatasetCore` actually works?

File: src/common/verify/isValidAccessGrant.ts (hunk 2 above)
Comment (garciafdezpatricia) [reply src/common/verify/isValidAccessGrant.ts:65]: Following the prior review, this change in the validations would be in `toVcDataset`?

Comment (NSeydoux) [reply src/common/verify/isValidAccessGrant.ts:65]: Absolutely

Comment (garciafdezpatricia) [reply src/common/verify/isValidAccessGrant.ts:65]: So, just to sum up, `vcToDataset` will either fetch the vc if it is a valid vc, that is:

`    if (typeof vc === "string" || vc instanceof URL) {
        return await getVerifiableCredential(vc.toString(), {
            returnLegacyJsonld: false,
            skipValidation: true,
            fetch: options.fetch,
        });
    }`

or, if it is not, convert it to a valid vc:

`if (
        typeof vc === "object" && 
        !isDatasetCore(vc) && 
        typeof (vc as { id: string }).id === "string"
    ) {
        return await verifiableCredentialToDataset(vc, {
            includeVcProperties: true,
            requireId: true,
        });
    }`

File: src/gConsent/manage/approveAccessRequest.ts [hunk 6]
```diff
@@ -287,11 +289,17 @@ export async function approveAccessRequest(
...
+  if (validVc === undefined && requestVc && isUrl(requestVc.toString())) {
+    validVc = requestVc;
+  }
 
   const internalGrantOptions = initializeGrantParameters(
-    typeof requestVc !== "undefined"
+    typeof validVc !== "undefined"
       ? await getBaseAccess(
-          requestVc,
+          validVc as DatasetWithId,
```
Comment (NSeydoux) [inline src/gConsent/manage/approveAccessRequest.ts:302]:
```suggestion
          validVc,
```
ditto

File: src/gConsent/manage/denyAccessRequest.ts [hunk 7]
```diff
@@ -91,8 +93,21 @@ async function denyAccessRequest(
     returnLegacyJsonld?: boolean;
   },
 ): Promise<DatasetWithId> {
+  let validVc;
+  validVc = await toVcDataset(vc, options);
```
Comment (NSeydoux) [inline src/gConsent/manage/denyAccessRequest.ts:97]:
```suggestion
  const validVc = await toVcDataset(vc, options);
```

File: src/gConsent/manage/denyAccessRequest.ts [hunk 8]
```diff
@@ -91,8 +93,21 @@ async function denyAccessRequest(
...
+
+  if (validVc === undefined && isUrl(vc.toString())) {
+    validVc = vc;
+  }
```
Comment (NSeydoux) [inline src/gConsent/manage/denyAccessRequest.ts:101]:
```suggestion
```

File: src/gConsent/manage/denyAccessRequest.ts [hunk 9]
```diff
@@ -91,8 +93,21 @@ async function denyAccessRequest(
...
+
+  if (validVc === undefined) {
+    throw new Error(
+      `Invalid argument: expected either a VC URL or a RDFJS DatasetCore, received ${vc}`,
+    );
+  }
+
   const baseVc: DatasetWithId = await getBaseAccess(
-    vc,
+    validVc as DatasetWithId,
```
Comment (NSeydoux) [inline src/gConsent/manage/denyAccessRequest.ts:110]:
```suggestion
    validVc,
```

File: src/gConsent/manage/revokeAccessGrant.ts [hunk 10]
```diff
@@ -53,7 +55,24 @@ async function revokeAccessCredential(
   types: NamedNode<string>[],
   options: Omit<AccessBaseOptions, "accessEndpoint"> = {},
 ): Promise<void> {
-  const credential = await getBaseAccess(vc, options, types[0]);
+  let validVc;
+  validVc = await toVcDataset(vc, options);
```
Comment (NSeydoux) [inline src/gConsent/manage/revokeAccessGrant.ts:59]:
```suggestion
  const validVc = await toVcDataset(vc, options);
```

File: src/gConsent/manage/revokeAccessGrant.ts [hunk 11]
```diff
@@ -53,7 +55,24 @@ async function revokeAccessCredential(
...
+
+  if (validVc === undefined && isUrl(vc.toString())) {
+    validVc = vc;
+  }
```
Comment (NSeydoux) [inline src/gConsent/manage/revokeAccessGrant.ts:63]:
```suggestion
```

File: src/gConsent/manage/revokeAccessGrant.ts [hunk 12]
```diff
@@ -53,7 +55,24 @@ async function revokeAccessCredential(
...
+
+  if (validVc === undefined) {
+    throw new Error(
+      `Invalid argument: expected either a VC URL or a RDFJS DatasetCore, received ${vc}`,
+    );
+  }
+
+  const credential = await getBaseAccess(
+    validVc as DatasetWithId,
```
Comment (NSeydoux) [inline src/gConsent/manage/revokeAccessGrant.ts:72]:
```suggestion
    validVc,
```

**Review by NSeydoux — CHANGES_REQUESTED** [review]

File: src/common/verify/isValidAccessGrant.test.ts [hunk 13]
```diff
@@ -112,6 +112,21 @@ describe("isValidAccessGrant", () => {
...
   });
 
+  it("uses the provided fetch if any from plain JSON VC", async () => {
+    jest.mocked(isVerifiableCredential).mockReturnValueOnce(true);
+    const mockedFetch = jest.fn<typeof fetch>().mockResolvedValueOnce(
+      new Response(JSON.stringify(MOCK_VERIFY_RESPONSE), {
+        status: 200,
+      }),
+    );
+    await isValidAccessGrant(JSON.parse(JSON.stringify(MOCK_ACCESS_GRANT)), {
```
Comment (NSeydoux) [inline src/common/verify/isValidAccessGrant.test.ts:122]:
```suggestion
    await isValidAccessGrant(MOCK_ACCESS_GRANT_BASE), {
```
Although what you had isn't incorrect, since `MOCK_ACCESS_GRANT` is actually derived from `MOCK_ACCESS_GRANT_BASE`, so we might as well use it directly.

File: src/common/verify/isValidAccessGrant.test.ts [hunk 14]
```diff
@@ -112,6 +112,21 @@ describe("isValidAccessGrant", () => {
     expect(mockedFetch).toHaveBeenCalled();
   });
 
+  it("uses the provided fetch if any from plain JSON VC", async () => {
```
Comment (NSeydoux) [inline src/common/verify/isValidAccessGrant.test.ts:115]: Unfortunately, this test doesn't really assert anything on the value passed to the verify endpoint, so it's not a great candidate to test for the plain JSON VC. That's on me, there wasn't a good way for you to know that in advance.

I think you can remove this test altogether, as the other one in this file covers the same use-case, but in a more meaningful way.

File: src/common/verify/isValidAccessGrant.test.ts [hunk 15]
```diff
@@ -137,6 +152,31 @@ describe("isValidAccessGrant", () => {
...
+      specCompliant: {},
+      legacy: {},
+    });
+    const mockedFetch = jest.fn<typeof fetch>().mockResolvedValueOnce(
+      new Response(JSON.stringify(MOCK_VERIFY_RESPONSE), {
+        status: 200,
+      }),
+    );
+
+    await isValidAccessGrant(MOCK_ACCESS_GRANT, {
```
Comment (NSeydoux) [inline src/common/verify/isValidAccessGrant.test.ts:168]:
```suggestion
    await isValidAccessGrant(MOCK_ACCESS_GRANT_BASE, {
```

File: src/common/verify/isValidAccessGrant.test.ts [hunk 16]
```diff
@@ -184,6 +248,32 @@ describe("isValidAccessGrant", () => {
...
+    await isValidAccessGrant(
+      JSON.parse(JSON.stringify("https://example.com/someVc")),
+      {
+        fetch: mockedFetch,
+        verificationEndpoint: MOCK_ACCESS_ENDPOINT,
+      },
+    );
+
+    expect(mockedFetch).toHaveBeenCalledWith("https://example.com/someVc");
+  });
```
Comment (NSeydoux) [inline src/common/verify/isValidAccessGrant.test.ts:275]:
```suggestion
```
In this test, it's not an Access Grant that is passed, but a URL, so we are already fetching and parsing the VC as we see fit. There is nothing to add to the original test.

File: src/common/verify/isValidAccessGrant.test.ts [hunk 17]
```diff
@@ -202,6 +292,27 @@ describe("isValidAccessGrant", () => {
...
+        JSON.parse(JSON.stringify("https://example.com/someVc")),
+        {
+          fetch: mockedFetch,
+          verificationEndpoint: MOCK_ACCESS_ENDPOINT,
+        },
+      ),
+    ).rejects.toThrow(
+      "Verifiable credential is not an object, or does not have an id",
+    );
+  });
```
Comment (NSeydoux) [inline src/common/verify/isValidAccessGrant.test.ts:314]:
```suggestion
```
Same as before, this is for a URL to dereference a VC, not a VC payload.

File: src/gConsent/manage/approveAccessRequest.ts [hunk 18]
```diff
@@ -22,10 +22,10 @@
...
 import { acp_ess_2 } from "@inrupt/solid-client";
-import type {
-  DatasetWithId,
-  VerifiableCredential,
-  VerifiableCredentialBase,
+import {
+  type DatasetWithId,
+  type VerifiableCredential,
+  type VerifiableCredentialBase,
 } from "@inrupt/solid-client-vc";
```
[the remark this answers is not in the document]
Comment (garciafdezpatricia) [reply src/gConsent/manage/approveAccessRequest.ts:29]: I'd say that would be the automatic imports, sorry about that

File: src/common/verify/isValidAccessGrant.test.ts [hunk 19]
```diff
@@ -227,6 +338,31 @@ describe("isValidAccessGrant", () => {
...
+      vcModule,
+      "getVerifiableCredentialApiConfiguration",
+    );
+    jest.mocked(isVerifiableCredential).mockReturnValueOnce(true);
+    const mockedFetch = jest.fn<typeof fetch>().mockResolvedValueOnce(
+      new Response(JSON.stringify(MOCK_VERIFY_RESPONSE), {
+        status: 200,
+      }),
+    );
+    await isValidAccessGrant(JSON.parse(JSON.stringify(MOCK_ACCESS_GRANT)), {
```
Comment (NSeydoux) [inline src/common/verify/isValidAccessGrant.test.ts:355]:
```suggestion
    await isValidAccessGrant(MOCK_ACCESS_GRANT_BASE), {
```

File: src/common/verify/isValidAccessGrant.test.ts [hunk 20]
```diff
@@ -256,6 +392,35 @@ describe("isValidAccessGrant", () => {
...
+        new Response(JSON.stringify(MOCK_ACCESS_GRANT), { status: 200 }),
+      )
+      // Then, the verification endpoint is called
+      .mockResolvedValueOnce(
+        new Response(JSON.stringify(MOCK_VERIFY_RESPONSE), {
+          status: 200,
+        }),
+      );
+
+    await isValidAccessGrant(JSON.parse(JSON.stringify(MOCK_ACCESS_GRANT)), {
```
Comment (NSeydoux) [inline src/common/verify/isValidAccessGrant.test.ts:417]:
```suggestion
    await isValidAccessGrant(MOCK_ACCESS_GRANT_BASE), {
```

File: src/common/verify/isValidAccessGrant.test.ts (hunk 15 above)
Comment (garciafdezpatricia) [reply src/common/verify/isValidAccessGrant.test.ts:168]: I tried with that, but since `isValidAccessGrant` input allows types are either `string`, `VerifiableCredential`, `DatasetWithId` or `URL`, with `MOCK_ACCESS_GRANT_BASE` being just an object, it raises a TS error.

File: src/gConsent/manage/revokeAccessGrant.test.ts [hunk 21]
```diff
@@ -112,6 +136,20 @@ describe("revokeAccessGrant", () => {
...
+        status: 401,
+        statusText: "Unauthorized",
+      }),
+    );
+    await expect(
+      revokeAccessGrant(JSON.parse(JSON.stringify("https://some.credential")), {
+        fetch: mockedFetch,
+      }),
+    ).rejects.toThrow(/\[https:\/\/some.credential\].*401.*Unauthorized/);
+  });
```
Comment (NSeydoux) [inline src/gConsent/manage/revokeAccessGrant.test.ts:151]:
```suggestion
```
Ditto

File: src/gConsent/manage/revokeAccessGrant.test.ts [hunk 22]
```diff
@@ -142,4 +190,29 @@ describe("revokeAccessGrant", () => {
...
+        fetch: mockedFetch,
+      },
+    );
+    expect(spiedRevoke).toHaveBeenCalledWith(
+      "https://some.issuer/status",
+      MOCKED_CREDENTIAL_ID,
+      expect.anything(),
+    );
+    expect(mockedFetch).not.toHaveBeenCalled();
+  });
```
Comment (NSeydoux) [inline src/gConsent/manage/revokeAccessGrant.test.ts:217]:
```suggestion
```
Ditto

File: src/common/verify/isValidAccessGrant.test.ts (hunk 15 above)
Comment (NSeydoux) [reply src/common/verify/isValidAccessGrant.test.ts:153]: Oh I see, then stringify and parse is fine, or you can do `as any`, which is equivalent. ESlint will complain about it, you can add an ignore mentioning we are explicitly testing for a behavior that is not compliant with our API type definitions.

File: src/gConsent/discover/redirectToAccessManagementUi.ts [hunk 23]
```diff
@@ -20,9 +20,9 @@
 //
 
 import type { UrlString, WebId } from "@inrupt/solid-client";
-import type {
-  DatasetWithId,
-  VerifiableCredential,
+import {
+  type DatasetWithId,
+  type VerifiableCredential,
```
Comment (NSeydoux) [inline src/gConsent/discover/redirectToAccessManagementUi.ts:25]: Is this change intentional, or is it introduced by `prettier`?

**Review by NSeydoux — APPROVED** [review]

File: src/gConsent/discover/redirectToAccessManagementUi.ts (hunk 23 above)
Comment (garciafdezpatricia) [reply src/gConsent/discover/redirectToAccessManagementUi.ts:25]: It wasn't intentional at all. I believe it has been introduced by `prettier` indeed
